File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/views/ViewGif.java
package com.sup.dev.android.views.views;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Movie;
import android.net.Uri;
import android.util.AttributeSet;
import android.view.View;

import com.sup.dev.android.R;
import com.sup.dev.android.app.SupAndroid;
import com.sup.dev.java.classes.callbacks.simple.Callback;
import com.sup.dev.java.libs.debug.Debug;
import com.sup.dev.java.tools.ToolsThreads;

import java.io.ByteArrayInputStream;
import java.io.FileNotFoundException;

public class ViewGif extends View {

    private static final int DEFAULT_MOVIE_VIEW_DURATION = 1000;

    private Movie movie;
    private long movieStart;
    private int currentAnimationTime;
    private float left;
    private float top;
    private float scale;
    private int measuredMovieWidth;
    private int measuredMovieHeight;
    private volatile boolean paused;
    private byte[] gifBytes;

    private Callback onGifLoaded;


    public ViewGif(Context context, AttributeSet attrs) {
        super(context, attrs);
        setLayerType(View.LAYER_TYPE_SOFTWARE, null);

        final TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.ViewGif, 0, R.style.Widget_GifView);
        int movieResourceId = array.getResourceId(R.styleable.ViewGif_android_src, 0);
        array.recycle();

        if (movieResourceId != 0)
            movie = Movie.decodeStream(getResources().openRawResource(movieResourceId));
    }


    //
    //  Methods
    //

    public void setGif(byte[] gifBytes) {
        this.gifBytes = gifBytes;
        Callback onGifLoaded = this.onGifLoaded;
        ToolsThreads.thread(() -> {
            movie = Movie.decodeStream(new ByteArrayInputStream(gifBytes));
            ToolsThreads.main(() -> {
                if(onGifLoaded != this.onGifLoaded)return;
                requestLayout();
                if (onGifLoaded != null) onGifLoaded.callback();
            });
        });
    }

    public void clear() {
        movie = null;
        onGifLoaded = null;
        requestLayout();
    }


    public void play() {
        if (this.paused) {
            this.paused = false;
            movieStart = android.os.SystemClock.uptimeMillis() - currentAnimationTime;
            invalidate();
        }
    }

    public void pause() {
        if (!this.paused) {
            this.paused = true;

            invalidate();
        }

    }

    private void updateAnimationTime() {
        long now = android.os.SystemClock.uptimeMillis();

        if (movieStart == 0) {
            movieStart = now;
        }

        int dur = movie.duration();

        if (dur == 0) {
            dur = DEFAULT_MOVIE_VIEW_DURATION;
        }

        currentAnimationTime = (int) ((now - movieStart) % dur);
    }

    private void drawMovieFrame(Canvas canvas) {

        movie.setTime(currentAnimationTime);

        canvas.save();
        canvas.scale(scale, scale);
        movie.draw(canvas, left / scale, top / scale);
        canvas.restore();
    }

    //
    //  Events
    //


    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {

        if (movie == null) {
            setMeasuredDimension(getSuggestedMinimumWidth(), getSuggestedMinimumHeight());
            return;
        }

        if(MeasureSpec.getSize(widthMeasureSpec) == 0 || MeasureSpec.getSize(heightMeasureSpec) == 0 || MeasureSpec.getMode(heightMeasureSpec) == MeasureSpec.UNSPECIFIED)
            scale = Math.max((float) MeasureSpec.getSize(widthMeasureSpec) / (float) movie.width(), (float) MeasureSpec.getSize(heightMeasureSpec) / (float) movie.height());
        else
            scale = Math.min((float) MeasureSpec.getSize(widthMeasureSpec) / (float) movie.width(), (float) MeasureSpec.getSize(heightMeasureSpec) / (float) movie.height());

        measuredMovieWidth = (int) (movie.width() * scale);
        measuredMovieHeight = (int) (movie.height() * scale);

        setMeasuredDimension(measuredMovieWidth, measuredMovieHeight);
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        super.onLayout(changed, l, t, r, b);
        left = (getWidth() - measuredMovieWidth) / 2f;
        top = (getHeight() - measuredMovieHeight) / 2f;
        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        if (movie != null) {
            if (!paused) {
                updateAnimationTime();
                drawMovieFrame(canvas);
                invalidate();
            } else {
                drawMovieFrame(canvas);
            }
        }
    }

    @Override
    public void onScreenStateChanged(int screenState) {
        super.onScreenStateChanged(screenState);
        invalidate();
    }

    @Override
    protected void onVisibilityChanged(View changedView, int visibility) {
        super.onVisibilityChanged(changedView, visibility);
        invalidate();
    }

    @Override
    protected void onWindowVisibilityChanged(int visibility) {
        super.onWindowVisibilityChanged(visibility);
        invalidate();
    }

    //
    //  Setters
    //

    public void setOnGifLoaded(Callback onGifLoaded) {
        this.onGifLoaded = onGifLoaded;
    }

    //
    //  Getters
    //


    public byte[] getGifBytes() {
        return gifBytes;
    }

    public boolean isPaused() {
        return this.paused;
    }

    public boolean isPlaying() {
        return !this.paused;
    }

}

File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/views/ViewGif.kt
package com.sup.dev.android.views.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Movie
import android.util.AttributeSet
import android.view.View
import com.sup.dev.android.R
import com.sup.dev.java.tools.ToolsThreads
import java.io.ByteArrayInputStream


class ViewGif(context: Context, attrs: AttributeSet) : View(context, attrs) {

    private var movie: Movie? = null
    private var movieStart: Long = 0
    private var currentAnimationTime: Int = 0
    private var left: Float = 0.toFloat()
    private var top: Float = 0.toFloat()
    private var scale: Float = 0.toFloat()
    private var measuredMovieWidth: Int = 0
    private var measuredMovieHeight: Int = 0
    @Volatile
    var isPaused: Boolean = false
        private set
    //
    //  Getters
    //


    var gifBytes: ByteArray? = null
        private set

    private var onGifLoaded: (() -> Unit)? = null

    val isPlaying: Boolean
        get() = !this.isPaused


    init {
        setLayerType(View.LAYER_TYPE_SOFTWARE, null)

        val array = context.obtainStyledAttributes(attrs, R.styleable.ViewGif, 0, R.style.Widget_GifView)
        val movieResourceId = array.getResourceId(R.styleable.ViewGif_android_src, 0)
        array.recycle()

        if (movieResourceId != 0)
            movie = Movie.decodeStream(resources.openRawResource(movieResourceId))
    }


    //
    //  Methods
    //

    fun setGif(gifBytes: ByteArray) {
        this.gifBytes = gifBytes
        val onGifLoaded = this.onGifLoaded
        ToolsThreads.thread {
            movie = Movie.decodeStream(ByteArrayInputStream(gifBytes))
            ToolsThreads.main {
                if (onGifLoaded === this.onGifLoaded) {
                    requestLayout()
                    if (onGifLoaded != null) onGifLoaded.invoke()
                }
            }
        }
    }

    fun clear() {
        movie = null
        onGifLoaded = null
        requestLayout()
    }


    fun play() {
        if (this.isPaused) {
            this.isPaused = false
            movieStart = android.os.SystemClock.uptimeMillis() - currentAnimationTime
            invalidate()
        }
    }

    fun pause() {
        if (!this.isPaused) {
            this.isPaused = true

            invalidate()
        }

    }

    private fun updateAnimationTime() {
        val now = android.os.SystemClock.uptimeMillis()

        if (movieStart == 0L) {
            movieStart = now
        }

        var dur = movie!!.duration()

        if (dur == 0) {
            dur = DEFAULT_MOVIE_VIEW_DURATION
        }

        currentAnimationTime = ((now - movieStart) % dur).toInt()
    }

    private fun drawMovieFrame(canvas: Canvas) {

        movie!!.setTime(currentAnimationTime)

        canvas.save()
        canvas.scale(scale, scale)
        movie!!.draw(canvas, left / scale, top / scale)
        canvas.restore()
    }

    //
    //  Events
    //


    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {

        if (movie == null) {
            setMeasuredDimension(suggestedMinimumWidth, suggestedMinimumHeight)
            return
        }

        val arg1 = View.MeasureSpec.getSize(widthMeasureSpec).toFloat() / movie!!.width().toFloat()
        val arg2 = View.MeasureSpec.getSize(heightMeasureSpec).toFloat() / movie!!.height().toFloat()

        if (View.MeasureSpec.getSize(widthMeasureSpec) == 0 || View.MeasureSpec.getSize(heightMeasureSpec) == 0
                || View.MeasureSpec.getMode(heightMeasureSpec) == View.MeasureSpec.UNSPECIFIED && movie!!.height() * Math.max(arg1, arg2) > View.MeasureSpec.getSize(heightMeasureSpec))
            scale = Math.max(arg1, arg2)
        else
            scale = Math.min(arg1, arg2)

        measuredMovieWidth = (movie!!.width() * scale).toInt()
        measuredMovieHeight = (movie!!.height() * scale).toInt()

        setMeasuredDimension(measuredMovieWidth, measuredMovieHeight)
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        super.onLayout(changed, l, t, r, b)
        left = (width - measuredMovieWidth) / 2f
        top = (height - measuredMovieHeight) / 2f
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        if (movie != null) {
            if (!isPaused) {
                updateAnimationTime()
                drawMovieFrame(canvas)
                invalidate()
            } else {
                drawMovieFrame(canvas)
            }
        }
    }

    override fun onScreenStateChanged(screenState: Int) {
        super.onScreenStateChanged(screenState)
        invalidate()
    }

    override fun onVisibilityChanged(changedView: View, visibility: Int) {
        super.onVisibilityChanged(changedView, visibility)
        invalidate()
    }

    override fun onWindowVisibilityChanged(visibility: Int) {
        super.onWindowVisibilityChanged(visibility)
        invalidate()
    }

    //
    //  Setters
    //

    fun setOnGifLoaded(onGifLoaded: ()->Unit) {
        this.onGifLoaded = onGifLoaded
    }

    companion object {

        private val DEFAULT_MOVIE_VIEW_DURATION = 1000
    }

}